### PR TITLE
Add -Zmetrics-dir=PATH to save diagnostic metadata to disk

### PR DESCRIPTION
**Pull request by yaahc**
Part of the Metrics Initiative: #128914

r? @estebank

### CODE REVIEW

File: compiler/rustc_driver_impl/src/lib.rs [hunk 1]
```diff
@@ -1307,15 +1310,14 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
...
+fn ice_path_with_config(config: Option<&UnstableOptions>) -> &'static Option<PathBuf> {
     ICE_PATH.get_or_init(|| {
         if !rustc_feature::UnstableFeatures::from_environment(None).is_nightly_build() {
             return None;
         }
-        if let Some(s) = std::env::var_os("RUST_BACKTRACE")
-            && s == "0"
-        {
-            return None;
-        }
```
Comment (yaahc) [inline compiler/rustc_driver_impl/src/lib.rs:1318]: this is unrelated to the rest of the feature. I've removed it because it was causing unintentional behavior (disabling backtraces were completely disabling saving ICEs, instead of just removing the backtrace output from the ICE report)

**Review by estebank — APPROVED** [review]
Only comment I have is to extend the make test to

 - exercise the new nightly flag
 - verify what happens when both the nightly flag and env variable are set
 - test the RUST_BACKTRACE=0 behavior against the file creation

File: compiler/rustc_driver_impl/src/lib.rs [hunk 2]
```diff
@@ -1307,15 +1310,14 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
 static ICE_PATH: OnceLock<Option<PathBuf>> = OnceLock::new();
 
 fn ice_path() -> &'static Option<PathBuf> {
+    ice_path_with_config(None)
```
Comment (compiler-errors) [inline compiler/rustc_driver_impl/src/lib.rs:1313]: Doesn't this result in the behavior that calling `ice_path()` *then* calling `ice_path_with_config(Some(/*...*/))` will return the ICE path w/ no config? 

Could you leave some sort of comment saying that it's intended that `ice_path_with_config` has been called first, but this will still return a path in case an ICE is emitted before the session has been initialized?

File: compiler/rustc_driver_impl/src/lib.rs [hunk 3]
```diff
@@ -1307,15 +1310,14 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
...
+}
+
+fn ice_path_with_config(config: Option<&UnstableOptions>) -> &'static Option<PathBuf> {
```
Comment (compiler-errors) [inline compiler/rustc_driver_impl/src/lib.rs:1316]: It would also perhaps be useful if we could do something like:

```
if ICE_PATH.get().is_some() && config.is_some() && cfg!(debug_assertions) {
    warn!("ICE_PATH has already been initialized -- files may be emitted at unintended paths")
}
```

File: compiler/rustc_driver_impl/src/lib.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_driver_impl/src/lib.rs:1313]: Alternatively, perhaps it's worthwhile to just parse the relevant env arg directly in `ice_path`?

**Review by compiler-errors — COMMENTED** [review]
I've got a comment about this initialization procedure's interaction with pre-session-init ICEs.

File: compiler/rustc_driver_impl/src/lib.rs (hunk 2 above)
Comment (yaahc) [reply compiler/rustc_driver_impl/src/lib.rs:1313]: The relevant configuration is a compiler flag, not an environment variable, unfortunately :(

And yeah, if you call the `ice_path()` function first it would mess up the `_with_config(Some(..))` version. I wrote this with the assumption that `ice_path()` would only ever be called from the ice hook, so there's no chance of a subsequent call to `_with_config`. 

I can definitely add comments clarifying this.

Comment (yaahc) [reply compiler/rustc_driver_impl/src/lib.rs:1328]: comment added, resolve if you're happy with it

File: tests/run-make/dump-ice-to-disk/rmake.rs [hunk 4]
```diff
@@ -42,23 +72,49 @@ fn main() {
...
-    assert!(content.contains("stack backtrace:"));
+fn test_flag_only(baseline: usize) {
+    clear_ice_files();
+    let metrics_arg = format!("-Zerror-metrics={}", cwd().display());
+    rustc().input("lib.rs").arg("-Ztreat-err-as-bug=1").arg(metrics_arg).run_fail();
+    let output = get_text_from_ice(cwd()).lines().count();
+    assert_eq!(output, baseline);
+}
+
+fn test_flag_and_env(baseline: usize) {
```
Comment (yaahc) [inline tests/run-make/dump-ice-to-disk/rmake.rs:105]: TODO

- [ ] warn when both are set

File: tests/run-make/dump-ice-to-disk/rmake.rs [hunk 5]
```diff
@@ -42,23 +72,49 @@ fn main() {
...
+    clear_ice_files();
+    let metrics_arg = format!("-Zerror-metrics={}", cwd().display());
+    let real_dir = cwd().join("actually_put_ice_here");
+    std::fs::create_dir(real_dir.clone()).unwrap();
```
Comment (jieyouxu) [inline tests/run-make/dump-ice-to-disk/rmake.rs:109]: Suggestion: we can use `rfs::create_dir` here (fs wrapper).

**Review by jieyouxu — APPROVED** [review]
Test changes looks good

File: compiler/rustc_session/src/options.rs [hunk 6]
```diff
@@ -1714,6 +1714,8 @@ options! {
         "emit the bc module with thin LTO info (default: yes)"),
     enforce_type_length_limit: bool = (false, parse_bool, [TRACKED],
         "enforce the type length limit when monomorphizing instances in codegen"),
+    error_metrics: Option<PathBuf> = (None, parse_opt_pathbuf, [UNTRACKED],
```
Comment (estebank) [inline compiler/rustc_session/src/options.rs:1717]: Lets rename this flag to something more general, like just `metrics` or `metrics_directory`, as it will be used for more things beyond "errors".

**Review by estebank — APPROVED** [review]
r=me after flag rename to `metrics-dir`
